app/v2/notifications/post_notifications.py

```diff
@@ -377,3 +435,88 @@ def get_reply_to_text(notification_type, form, template):
 
 def strip_keys_from_personalisation_if_send_attach(personalisation):
     return {k: v for (k, v) in personalisation.items() if not (type(v) is dict and v.get("sending_method") == "attach")}
+
+
+def check_for_csv_errors(recipient_csv, max_rows, remaining_messages):
+    nb_rows = len(recipient_csv)
+
+    if recipient_csv.has_errors:
+        if recipient_csv.missing_column_headers:
+            raise BadRequestError(
+                message=f"Missing column headers: {', '.join(sorted(recipient_csv.missing_column_headers))}",
+                status_code=400,
+            )
+        if recipient_csv.duplicate_recipient_column_headers:
+            raise BadRequestError(
+                message=f"Duplicate column headers: {', '.join(sorted(recipient_csv.duplicate_recipient_column_headers))}",
+                status_code=400,
+            )
+        if recipient_csv.more_rows_than_can_send:
+            raise BadRequestError(
+                message=f"You only have {remaining_messages} remaining messages before you reach your daily limit. You've tried to send {nb_rows} messages.",
+                status_code=400,
+            )
+
+        if recipient_csv.too_many_rows:
+            raise BadRequestError(
+                message=f"Too many rows. Maximum number of rows allowed is {max_rows}",
+                status_code=400,
+            )
+        if not recipient_csv.allowed_to_send_to:
+            if api_user.key_type == KEY_TYPE_TEAM:
+                explanation = "because you used a team and safelist API key."
+            if authenticated_service.restricted:
+                explanation = (
+                    "because your service is in trial mode. You can only send to members of your team and your safelist."
+                )
+            raise BadRequestError(
+                message=f"You cannot send to these recipients {explanation}",
+                status_code=400,
+            )
+        if recipient_csv.rows_with_errors:
+
+            def row_error(row):
+                content = []
+                for header in [header for header in recipient_csv.column_headers if row[header].error]:
+                    if row[header].recipient_error:
+                        content.append(f"`{header}`: invalid recipient")
+                    else:
+                        content.append(f"`{header}`: {row[header].error}")
+                return f"Row {row.index} - {','.join(content)}"
+
+            errors = ". ".join([row_error(row) for row in recipient_csv.initial_rows_with_errors])
+            raise BadRequestError(
+                message=f"Some rows have errors. {errors}.",
+                status_code=400,
+            )
+        else:
+            raise NotImplementedError("Got errors but code did not handle")
+
+
+def create_bulk_job(service, api_key, template, form, recipient_csv):
```

Comment:
After the request has been validated, upload the built CSV payload to S3, create a job in the database. Dispatch a Celery task now if the job is not scheduled to start processing immediately. Otherwise it will be started later on.